core/base/batch_multi_vector.cpp

```diff
@@ -0,0 +1,256 @@
+        }
+    }
+    this->get_executor()->run(batch_multi_vector::make_scale(alpha, this));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::add_scaled_impl(
+    const BatchMultiVector<ValueType>* alpha,
+    const BatchMultiVector<ValueType>* b)
+{
+    GKO_ASSERT_EQ(alpha->get_num_batch_entries(),
+                  this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_ROWS(alpha->get_common_size(), dim<2>(1, 1));
+    for (size_type b = 0; b < alpha->get_num_batch_entries(); ++b) {
+        if (alpha->get_common_size()[1] != 1) {
+            // different alpha for each column
+            GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
+                                  alpha->get_common_size());
+        }
+    }
+    GKO_ASSERT_EQ(b->get_num_batch_entries(), this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(this->get_common_size(), b->get_common_size());
+
+    this->get_executor()->run(
+        batch_multi_vector::make_add_scaled(alpha, b, this));
+}
+
+
+inline const batch_dim<2> get_col_sizes(const batch_dim<2>& sizes)
+{
+    return batch_dim<2>(sizes.get_num_batch_entries(),
+                        dim<2>(1, sizes.get_common_size()[1]));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::compute_dot_impl(
+    const BatchMultiVector<ValueType>* b,
+    BatchMultiVector<ValueType>* result) const
+{
+    GKO_ASSERT_EQ(b->get_num_batch_entries(), this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(this->get_common_size(), b->get_common_size());
+    GKO_ASSERT_EQ(this->get_num_batch_entries(),
+                  result->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(
+        result->get_common_size(),
+        get_col_sizes(this->get_size()).get_common_size());
+    this->get_executor()->run(
+        batch_multi_vector::make_compute_dot(this, b, result));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::compute_norm2_impl(
+    BatchMultiVector<remove_complex<ValueType>>* result) const
+{
+    GKO_ASSERT_EQ(this->get_num_batch_entries(),
+                  result->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(
+        result->get_common_size(),
+        get_col_sizes(this->get_size()).get_common_size());
+    this->get_executor()->run(batch_multi_vector::make_compute_norm2(
+        as<BatchMultiVector<ValueType>>(this), result));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::convert_to(
+    BatchMultiVector<next_precision<ValueType>>* result) const
+{
+    result->values_ = this->values_;
+    result->set_size(this->get_size());
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::move_to(
+    BatchMultiVector<next_precision<ValueType>>* result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename MatrixType, typename MatrixData>
+inline void read_impl(MatrixType* mtx, const std::vector<MatrixData>& data)
+{
+    auto common_size = data[0].size;
```

Comment:
This segfaults if data is empty!

Reply:
The GKO_ASSERT does nothing in release mode. So perhaps use `GKO_INVALID_STATE`, which will throws.